### ISSUE TITLE
nvfuserex has problems taking getitem. 

### ISSUE THREAD

**Issue by wujingyue** [Lightning-AI/lightning-thunder#482]
getitem_nvfuser tests failed for a similar reason to https://github.com/Lightning-AI/lightning-thunder/blob/54bb6146ff757905925f8d9ea2197870c4971011/thunder/tests/opinfos.py#L3113-L3115. I can again create a wrapper so `slice` objects don't get passed to FusionDefinitionWrapper. But I'd love to hear thoughts from @jjsjann123 and @kevinstephano who may have a better fix.

_Originally posted by @wujingyue in https://github.com/Lightning-AI/lightning-thunder/issues/191#issuecomment-2119408081_
            

cc @tfogal

**Comment by wujingyue** [comment]
To reproduce this, check out branch `wjy/bookend` and run `pytest thunder/tests/test_ops.py -k test_core_vs_torch_consistency_getitem_nvfuser_cuda_float32 -s`. 

The branch disables bookend (a workaround that we want to get rid of) so slice prims are given to nvfuser, adds some debug prints and highlights the failing test. 

The immediate error is 
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
thunder/tests/test_ops.py:76: in test_core_vs_torch_consistency
    result = run_snippet(
thunder/tests/framework.py:516: in run_snippet
    raise ex
thunder/tests/framework.py:508: in run_snippet
    snippet(*args, **kwargs)
thunder/tests/test_ops.py:37: in snippet_torch_consistency
    thunder_result = op(*sample.args, **sample.kwargs)
thunder/__init__.py:620: in fn_
    result = cache_entry.computation_fn(*inps)
/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py:115: in decorate_context
    return func(*args, **kwargs)
/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py:28: in decorate_autocast
    return func(*args, **kwargs)
/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py:28: in decorate_autocast
    return func(*args, **kwargs)
thunder.computation_1:10: in computation
    [t0] = nvFusion0(a, b)
thunder/executors/nvfuserex_impl.py:401: in __call__
    fd = self.get_fd(to_descriptors(args))
thunder/executors/nvfuserex_impl.py:384: in to_descriptors
    return tuple(to_descriptor(arg) for arg in args)
thunder/executors/nvfuserex_impl.py:384: in <genexpr>
    return tuple(to_descriptor(arg) for arg in args)
thunder/executors/nvfuserex_impl.py:372: in to_descriptor
    numbertype, _ = check_same_dtype(*arg)
thunder/core/utils.py:220: in check_same_dtype
    typ = to_dtype(a, true_dtype=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = slice(1, 3, 1)

    def to_dtype(x: Any, /, *, true_dtype: bool = False) -> None | dtype:
        """Exctracts a dtype from an object or class."""

        if x is None or isinstance(x, dtype):
            return x
        if isinstance(x, np.ndarray):
            return _numpy_to_thunder_dtype_map[x.dtype]
        if isinstance(x, np.dtype):
            return _numpy_to_thunder_dtype_map[x.type]
        if isinstance(x, torch.Tensor):
            return _torch_to_thunder_dtype_map[x.dtype]
        if isinstance(x, torch.dtype):
            return _torch_to_thunder_dtype_map[x]
        if isinstance(x, TensorProxyInterface):
            if true_dtype:
                return x.true_dtype
            return x.dtype
        if isinstance(x, dtype):
            return x
        if isinstance(x, NumberProxyInterface):
            return x.python_type
        if isinstance(x, Number):
            return _numberclass_to_numbertype(type(x))
        if isinstance(x, type) and issubclass(x, Number):
            return _numberclass_to_numbertype(x)

>       raise ValueError(f"Trying to extract a dtype from object {x} with unknown type {type(x)}!")
E       ValueError: Trying to extract a dtype from object slice(1, 3, 1) with unknown type <class 'slice'>!

thunder/core/dtypes.py:300: ValueError
```
This is triggered by to_descriptors which tries to compute a cache key for the function signature and fails on `slice`. 

However, even with that check removed, 
```
diff --git a/thunder/executors/nvfuserex_impl.py b/thunder/executors/nvfuserex_impl.py
index bc0708ef..e30bf428 100644
--- a/thunder/executors/nvfuserex_impl.py
+++ b/thunder/executors/nvfuserex_impl.py
@@ -368,13 +368,13 @@ def to_descriptors(args) -> tuple:
         if isinstance(arg, Number):
             return type(arg)
         elif isinstance(arg, tuple):
-            if len(arg) != 0:
-                numbertype, _ = check_same_dtype(*arg)
-                check(
-                    numbertype == int,
-                    lambda: f"tuple in nvfuser only supports int, but found {numbertype}",
-                    exception_type=AssertionError,
-                )
+            # if len(arg) != 0:
+            #     numbertype, _ = check_same_dtype(*arg)
+            #     check(
+            #         numbertype == int,
+            #         lambda: f"tuple in nvfuser only supports int, but found {numbertype}",
+            #         exception_type=AssertionError,
+            #     )
             return type(arg)
         elif isinstance(arg, torch.Tensor):
             return (*get_tensor_descriptor(arg), arg.dtype)
```
nvFuser still fails with the following RuntimeError: 
```
RuntimeError: Tracer cannot infer type of slice(1, 3, 1)
:Only tensors and (possibly nested) tuples of tensors, lists, or dictsare supported as inputs or outputs of traced functions, but instead got value of type slice.
```

**Comment by wujingyue** [comment]
FWIW, @jjsjann123 and I briefly looked at the Thunder trace before and after nvFusion. 

```
*** before nvFusion ***
# Constructed by Dead Code Elimination (took 0 milliseconds)
import thunder
import thunder.torch as ltorch
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, b):
  # a: "cuda:0 f32[5, 5]"
  # b: "tuple"
  t0 = ltorch.getitem(a, b)  # t0: "cuda:0 f32[2, 2]"
    # t0 = prims.slice_prim(a, [1, 2], [3, 4], [1, 1])  # t0: "cuda:0 f32[2, 2]"
  return t0

*** after nvFusion ***
# Constructed by Fusion (took 1 milliseconds)
import torch
from thunder.executors.torchex import no_autocast

@torch.no_grad()
@no_autocast
def computation(a, b):
  # a: "cuda:0 f32[5, 5]"
  # b: "tuple"
  [t0] = nvFusion0(a, b)
    # t0 = prims.slice_prim(a, [1, 2], [3, 4], [1, 1])  # t0: "cuda:0 f32[2, 2]"
  return t0
```

@jjsjann123 felt like there's something wrong with the trace to begin with. `b` isn't used any more and has been baked in as constant already.

**Comment by tfogal** [comment]
@jjsjann123 can take a first stab at debugging and then pass it off to someone.

triage team: 1) thoughts? 2) who do we consider the maintainer of the nvfuser executor? Ivan? 3) maybe same as (2), but who can we assign this to to drive a solution?

**Comment by jjsjann123** [comment]
This will be fixed by #544 . Thank a lot to @kiya00 for helping out.
